scalable-go-service/index.md

```diff
@@ -0,0 +1,322 @@
+contents of an individual request dictated how much internal resources were
+allocated. Even if the request was validated to ensure the underlying capacities
+were not abused, this solution did not give us the freedom to configure how much
+resources we wanted to make available.
+
+## The Worker Model
+
+The proof of concept gave us confidence that we were on the right path, so we
+decided to take the time to design a more robust concurrent model to address
+some of our concerns. We began by identifying all the components that could
+conceptually run independent of the others in the system. The white board gave
+us the freedom to compose these independently executing modules and find areas
+where we needed to add communication channels between them.
+
+<p align="center">
+    <img src="./pictures/second.jpg" width="600" heigth:"400"/>
+</p>
+
+The route handler, the token manager, and the rate loader were the three
+components we identified as required for processing a request to load rates. It
+was not immediately obvious how these components would run independently. The
+rate loader takes a validated request from the route handler and needs a token
+before contacting the vendor. However, we noticed that multiple rate loaders can
+execute independently once they have all of the necessary info. Once we
+facilitate proper communication between the components, we can start regarding
+them as workers that receive required parameters from `channels`, perform their
+task, and communicate the result through `channels`. Groups of workers that
+perform the same tasks can be regarded as a pool of workers. They all listen on
+a single channel for requests they can process.
+
+<p align="center">
+    <img src="./pictures/third.jpg" width="600" heigth:"400"/>
+</p>
+
+
+How many rate loaders? It depends on the application, but this is now a
+configurable parameter. It does not have to be an exact number either. One might
+configure a lower and an upper bound for how many rate loaders they want in
+their app and can add logic to dynamically scale to throw limits based on the
+current load.
+
+## Sample Code
+
+#### Main
+
+```go
+func main() {
+	var wg sync.WaitGroup
+	tokenRQChan := make(chan domain.TokenRQ)
+	ratesRQChan := make(chan domain.RateRQ, rateLoaderCount*2)
+
+	var tokenLoader = TokenLoader{
+		In:   tokenRQChan,
+	}
+	var rateLoaderPool = RateLoaderPool{
+		Count:         viper.GetInt("rateLoaderCount"),
+		In:            ratesRQChan,
+		TokenRQChan:   tokenRQChan,
+	}
+
+	appRouter := mux.NewRouter()
+	appRouter.HandleFunc("/rates", RatesHandler(ratesRQChan))
+	srv := http.Server{
+		Addr:    viper.GetString("server.host") + ":" + viper.GetString("server.port"),
+		Handler: appRouter,
+	}
+
+	wg.Add(1)
+	go func() {
+		tokenLoader.Start()
+		wg.Done()
+	}()
+
+	wg.Add(1)
+	go func() {
+		rateLoaderPool.Start()
+		wg.Done()
+	}()
+
+    // omitting implementation of gracefully shutting down all the workers
```

Comment:
you should at least include the `wg.Wait()` line. Otherwise this main function will exit immediately.